metricflow-semantics/tests_metricflow_semantics/time/test_time_adjuster.py

```diff
@@ -0,0 +1,156 @@
+                pandas_end_of_period = pandas_adjuster.adjust_to_end_of_period(time_granularity, date_time)
+            dateutil_start_of_period = dateutil_adjuster.adjust_to_start_of_period(time_granularity, date_time)
+            dateutil_end_of_period = dateutil_adjuster.adjust_to_end_of_period(time_granularity, date_time)
+            assert (
+                pandas_start_of_period or dateutil_start_of_period
+            ) == dateutil_start_of_period, f"start-of-period mismatch: {date_time.isoformat()} {time_granularity}"
+            assert (
+                pandas_end_of_period or dateutil_end_of_period
+            ) == dateutil_end_of_period, f"end-of-period mismatch: {date_time.isoformat()} {time_granularity}"
+            rows.append(
+                (
+                    date_time.isoformat(),
+                    time_granularity.name,
+                    dateutil_start_of_period.isoformat(),
+                    dateutil_end_of_period.isoformat(),
+                )
+            )
+    assert_str_snapshot_equal(
+        request=request,
+        mf_test_configuration=mf_test_configuration,
+        snapshot_id="results",
+        snapshot_str=tabulate.tabulate(rows, headers=["Date", "Grain", "Period Start", "Period End"]),
+    )
+
+
+def test_expand_time_constraint_to_fill_granularity(  # noqa: D103
+    date_times_to_check: Sequence[datetime.datetime], grain_to_count_in_year: Dict[TimeGranularity, int]
+) -> None:
+    pandas_adjuster = PandasTimePeriodAdjuster()
+    dateutil_adjuster = DateutilTimePeriodAdjuster()
+
+    test_cases = tuple(
+        (start_time, end_time, time_granularity)
+        for start_time in date_times_to_check
+        for time_granularity in TimeGranularity
+        for end_time in (
+            start_time + datetime.timedelta(days=day_offset)
+            for day_offset in range(grain_to_count_in_year[time_granularity] + 2)
+        )
+    )
+
+    test_case_count = len(test_cases)
+    logger.info(f"There are {test_case_count} test cases")
+
+    finished_count = 0
+
+    for start_time, end_time, time_granularity in test_cases:
+        time_constraint = TimeRangeConstraint(start_time=start_time, end_time=end_time)
+        pandas_adjuster_result = pandas_adjuster.expand_time_constraint_to_fill_granularity(
+            time_constraint, time_granularity
+        )
+
+        dateutil_adjuster_result = dateutil_adjuster.expand_time_constraint_to_fill_granularity(
+            time_constraint, time_granularity
+        )
+
+        assert (
+            pandas_adjuster_result == dateutil_adjuster_result
+        ), f"Expansion mismatch: {pandas_adjuster_result=} {dateutil_adjuster_result=} {time_granularity=}"
+        finished_count += 1
+        if finished_count % 100000 == 0 or finished_count == test_case_count:
+            logger.info(f"Progress {finished_count / test_case_count * 100:.0f}%")
```

Comment:
How many INFO rows does this print? If it's a lot we might want to put it in debug just so it's not filling up CI logs, but I assume it' snot that many since it's one every 100k.

Reply:
<10 lines